### PR TITLE
COO intersection primitives: performance improvement

### PR DESCRIPTION
**Pull request by nikitaved**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #95165
* #94991
* #94596
* #94406
* #94401
* __->__ #96094

cc @alexsamardzic @pearu @cpuhrsch @amjames @bhosmer

### CODE REVIEW

**Comment by nikitaved** [comment]
@cpuhrsch , @pearu , could you please stamp this one again?

File: aten/src/ATen/native/sparse/SparseBinaryOpIntersectionCommon.h [hunk 1]
```diff
@@ -617,33 +436,23 @@ void _sparse_binary_op_intersection_kernel_out(
 
   const auto broadcasted_shape = infer_size(x.sizes(), y.sizes());
 
-  int64_t max_hash_val = 1;
-  for (const auto d : c10::irange(x.sparse_dim())) {
-    max_hash_val *= broadcasted_shape[d];
+  // 8 sparse dims should be more than enough?
+  constexpr int64_t max_sparse_dims = 8;
+
```
Comment (nikitaved) [inline aten/src/ATen/native/sparse/SparseBinaryOpIntersectionCommon.h:441]: Convert to `size_t` but make sure no `size_t-int64` comparisons take place.

**Review by pearu — APPROVED** [review]
LGTM!

**Pull request closed without merging**